sdk/storage/Azure.Storage.Common/src/Shared/StorageClientOptions.cs

```diff
@@ -65,18 +66,21 @@ public static HttpPipelinePolicy AsPolicy<TUriBuilder>(this AzureSasCredential c
         /// Get an authentication policy to sign Storage requests.
         /// </summary>
         /// <param name="credential">Credential to use.</param>
+        /// <param name="options"> The <see cref="ISupportsTenantIdChallenges"/> to apply to the credential. </param>
         /// <returns>An authentication policy.</returns>
-        public static HttpPipelinePolicy AsPolicy(this TokenCredential credential) =>
-            new BearerTokenAuthenticationPolicy(
+        public static HttpPipelinePolicy AsPolicy(this TokenCredential credential, ClientOptions options) =>
+            new StorageBearerTokenChallengeAuthorizationPolicy(
                 credential ?? throw Errors.ArgumentNull(nameof(credential)),
-                StorageScope);
+                StorageScope,
+                options is ISupportsTenantIdChallenges { EnableTenantDiscovery: true });
```

Comment:
what does it do? It looks like always enabling tenant discovery.

Reply:
It's a pattern match for checking that the enable value is true.

Reply:
That's not easy to read I have to say. 
Does it allocate anonymous `ISupportsTenantIdChallenges` in order to do the matching or does compiler optimize it at some point? If it does allocate then I'd rather extract this logic and just type-check/cast to simplify for some readers.
Btw. we're doing it because Shares don't have this feature right ? Other option to consider is something like https://github.com/Azure/azure-sdk-for-net/blob/70a4fd814f0bf8938114044bc5cbd8e5ecc7e8f0/sdk/storage/Azure.Storage.Common/src/Shared/StorageVersionExtensions.cs#L13 .

Reply:
No, it [compiles to the same](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgAwFgBQaCMOOyeAnABQBEgPBuAg+xQJQDcRe6ABAGYD2XbAvGwrAApgGdgFZtgBuAQwBObAMb82AO2EB3NgGEyTHACMeENmFEAZYWoDmwABYAxLgFdFApQDoACrOAj5NTJuLgNscytbB2c3VS8AOS5ff2FA4J4wsA42MgjrOydXeXo2YnIKYnZzNgh8hzYAFgYpIgBmUtRdHABvHDZ+0vbjLlNkgKCAeUMAK2ElYDZQvoHe7AH10oB2RbNRUtY2bsiCthBGgF8pdfPCNYHkIZM2RLHUyZm5haW7/tWN9eQ2141UqbD2ADJwWDPMd6nwBA0rgMbthzkA=) as if no pattern was used.

Reply:
I don't think it does - [sharplab](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgAwFgBQaCMOOyeAnABQBEgPBuAg+xQJQDcRe6ABAGYD2XbAvGwrAApgGdgFZtgBuAQwBObAMb82AO2EB3NgGEyTHACMeENmFEAZYWoDmwABYAxLgFdFApQDoACrOAj5NTJuLgNscytbB2c3VS8AOS5ff2FA4J4wsA42MgjrOydXeXo2YnIKYnZzNgh8hzYAFgYpIgBmUtRdHABvHDZ+0vbjLlNkgKCAeUMAK2ElYDZQvoHe7AH10oB2RbNRUtY2bsiCthBGgF8pdfPCNYHkIZM2RLHUoMrF+mX+1Y315G2XE8x3qfAEDSuAxu2HOQA==)